platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,175 @@
+
+Let's assume a metric with the category `myCategory`, the name `myName`.
+As today the key of the metric is `myCategory.myName`.
+Next to that the metric will have 2 labels: `category=myCategory` and `name=myName`.
+The unique identifier of the metric is the key and the labels.
+That is a breaking change since today the key is the unique identifier of a metric.
+Since we do not have api today to add labels the migration should be minimalistic.
+
+#### Migration of the prometheus endpoint
+
+The `nodeId` of a platform is already added as a label today when using the prometheus endpoint.
+For that endpoint we just need to remove the hard coded `nodeId` label and add a general algorithm to add labels.
+
+#### Migration of the CSV exporter
+
+The CSV exporter use the `nodeId` today to create the file name.
+Since the CSV exporter is only used by the platform module we can continue with that approach.
+
+## Testing
+
+The metrics module is tested by unit tests today. We need to add tests for the new concept of labels.
+Next to that we need to add tests of the CSV exporter and the prometheus endpoint to ensure that the labels are added correctly.
+Here we need to check that a platform metric (that has a `nodeId` label) is still working correctly and ends in the exact same output as today.
+
+## Open Questions
+
+Please help me to answer the following questions before we can start with the implementation:
+
+- Should the `getLabels()` method return `Set`, `Collection` or `List`?
+- Should we add a new methods to the `Metrics` interface to get for example all metrics with a specific label?
+- How will a label with a `null` value be handled?
+
+## Alternatives
+
+Since the idea of this proposal is to make the metrics module independent of the platform module we need to remove
+the usage of the `NodeId` class or extract the class from the platform and define it as general API.
+Since the class does not make any sense outside the consensus node (like in block node for example) this is not a good idea.
+Another option would be to change the type of the `nodeId` to a `String` but that would end in an ugly API for any
+system next to the platform that uses the metrics module since `null` or any other default must always be used as the `nodeId`.
+Next to that all that ideas will end in having a hardcoded `nodeId` label in the metrics that will be forwarded to the monitoring system.
+
+## Future Work
+
+As a next step we can think about labels next to `nodeId` that makes sense to have in our metrics.
+For example, we can add a label called `type` that defines if a metric is a counter, gauge, or histogram.
+Another option is to add labels for layers (like `base`, `consensus`, `services`, ...).
+By doing so we can add more information to the metrics that can be used by the monitoring system.
+
+## References
+
+- https://prometheus.io/docs/practices/naming/#labels
+- https://prometheus.io/docs/practices/instrumentation/#use-labels
```

Comment:
From this article:
```
Use labels
Few monitoring systems have the notion of labels and an expression language to take advantage of them, so it takes a bit of getting used to.

When you have multiple metrics that you want to add/average/sum, they should usually be one metric with labels rather than multiple metrics.

For example, rather than http_responses_500_total and http_responses_403_total, create a single metric called http_responses_total with a code label for the HTTP response code. You can then process the entire metric as one in rules and graphs.
```

Reply:
I would certainly prefer the second option; labels can be appended, and combined in different ways and sometimes omitted. Also, metrics instances are sometimes used as instance fields:

```
class SomeClass{
  private Metric metric; 

  SomeClass(){
    metric =metrics.getOrCreate("http_metric"); 
  }


  int myHttpHandler(String methodName, String deviceType){

     int status = handle();
     List<Label> labels = 
       List.of(Label.of(methodName), Label.of(status), Label.of(deviceType)):
   
     metric.increase(labels);
     //or metric.increase();
     //or metric.increase(Label.of(methodName));
     //or metric.increase(Label.of(methodName), Label.of(deviceType));
    //or metric.increase(Label.of(status), Label.of(deviceType));
     return status;
  }
}
```

The first option has the drawback that:
a) The user must define a label for every possible `int` that `handle()` would return when creating the metric. This seems plausible with the status code example, but how about other situations where it could return anything?
b) you must change all instances where the metric is stored in a field.